library/arith.lisp

```diff
@@ -480,6 +493,136 @@ The fields are defined as follows:
+  (declare sincos (Trigonometric :a => :a -> (Tuple :a :a)))
+  (define (sincos x)
+    (tuple (sin x) (cos x)))
+
+  (define-class (Exponentiable :a)
+    (**      (:a -> :a -> :a))
+    (logBase (:a -> :a -> :a)))
+
+  (declare sqrt ((Dividable Integer :a) (Exponentiable :a) => :a -> :a))
+  (define (sqrt x)
+    "Returns the square root of the input"
+    (** x (general/ (the integer 1) (the integer 2))))
+
+  (define-class ((Num :a) (Dividable :a :a) (Trigonometric :a) (Exponentiable :a) => Float :a)
+    (ee :a)
+    (pi :a))
+
+  (declare exp ((Float :f) => :f -> :f))
+  (define (exp x)
+    "Raise ee to the power of X"
+    (** ee x))
+
+  (declare log ((Float :f) => :f -> :f))
+  (define (log x)
+    "The natural logarithm of a given value"
+    (logBase ee x))
+
+  (define-class ((Float :f) (Ord :f) => RealFloat :f)
+    "Real Floating-Point Numbers"
+    (rationalize (:f -> Fraction)))
+
+  (declare atan2 ((RealFloat :f) => :f -> :f -> :f))
+  (define (atan2 y x)
+    "computes the arc tangent of y/x,
+     using the signs to determine the return value's quadrant"
+    (match (tuple (<=> x 0) (<=> y 0))
+      ((tuple (GT) _)    (atan (/ y x)))
+      ((tuple (LT) (LT)) (- (atan (/ y x)) pi))
+      ((tuple (LT) _)    (+ (atan (/ y x)) pi))
+      ((tuple (EQ) (GT)) (/ pi (fromint 2)))
+      ((tuple (EQ) (LT)) (/ pi (fromint -2)))
+      ((tuple (EQ) (EQ)) 0))))
+
+(cl:eval-when (:compile-toplevel :load-toplevel :execute)
+  (cl:defun %define-unary-wrapper
+      (return-type function cl:&key
+       (package :cl)
+       (cl-fun  (cl:find-symbol (cl:symbol-name function) package))
+       domain)
+    "Defines a function which is a unary wrapper around a corresponding CL function.
+     Takes an optional domain, which is an boolean expression in terms of 'x',
+     which returns true iff if a value is in the domain"
+    `(define (,function x)
+       (lisp ,return-type (x)
+         ,(cl:when domain
+            `(cl:unless ,domain 
+               (cl:error "~a is not in the domain of ~a for ~a"
+                         x ',function ',return-type)))
+         (,cl-fun x)))))
+
+(cl:defmacro %define-float-instances
+    (coalton-type cl:&key
+     (package :cl)
+     (underlying-type (cl:find-symbol (cl:symbol-name coalton-type) package)))
+  "Defines the float instances for a lisp floating-point type"
+  `(coalton-toplevel
+     (define-instance (Trigonometric ,coalton-type)
+       ,(%define-unary-wrapper coalton-type 'sin   :package package)
+       ,(%define-unary-wrapper coalton-type 'cos   :package package)
+       ,(%define-unary-wrapper coalton-type 'tan   :package package)
+       ,(%define-unary-wrapper coalton-type 'sinh  :package package)
+       ,(%define-unary-wrapper coalton-type 'cosh  :package package)
+       ,(%define-unary-wrapper coalton-type 'tanh  :package package)
+       ,(%define-unary-wrapper coalton-type 'asin  :package package
+                                                   :domain '(cl:<= -1 x 1))
```

Comment:
this domain is only valid for real arguments; if you account for complex ones you'll have to define things separately